wireguard: Use dummy peer for allowed IP removal

jrife opened the pull request:
Recent flakes in the "Network performance GKE" tests, in particular those that use wireguard, uncovered an issue where updates to a peer's allowed IPs can cause connectivity issues. In the best case this may just result in dropped packets while in the worst case may lead to failed calls to `sendto`/`sendmsg` (see GH-33159 for more detail on some of the symptoms observed while running netperf).
    
This issue happens when `ReplaceAllowedIPs` is set, as it causes the wireguard driver to completely clear the set of allowed IPs for a peer before the set is rebuilt. This leaves a gap where packets sent or received at that time will appear to not have a peer associated with them. The current implementation sets `ReplaceAllowedIPs` every time `updatePeerByConfig` is called, but we can avoid ever needing to use it.

This commit uses a "dummy peer" as a way to remove allowed IPs from a peer rather than using `ReplaceAllowedIPs`. We use `peerConfig` to track any pending IP inserts and removals and split these into two different sets of updates in `updatePeerByConfig`. The first call to `ConfigureDevice()` adds any IPs that are pending insertion. Next, any IPs that are pending removal are moved to the "dummy peer" before removing the dummy peer altogether, the net effect being that the IP is removed from the original peer's list of allowed IPs.
    
This commit mostly sidesteps the issue that GH-34095 ("wireguard: restore allowedIPs upong agent restart") set out to solve since we only ever explicitly remove allowed IPs, meaning we can do away with a lot of the accounting previously done by `restoredPeers` and `restoredAllowedIPs`. However, we retain the behavior in `RestoreFinished()` that removes stale peers and allowed IPs.

## Testing
I performed the steps outlined in #33159 in a local kind cluster both before and after applying the change contained in this PR. I ran `netperf` manually while executing this script which induces failure by forcing lots of calls to `updatePeerByConfig`.

```bash
#!/bin/bash

count=100
for i in $(seq $count); do
	kubectl label --overwrite nodes kind-control-plane kill-netperf=true
	kubectl label --overwrite nodes kind-control-plane kill-netperf=false
done
```
### Before
`netperf` fails every time consistently within seconds.

```bash
jrife@jrife-kvm:~/code/cilium$ kubectl exec -it -n cilium-test-1 perf-client-other-node-749b79954c-hpgts -- bash
[root@perf-client-other-node-749b79954c-hpgts netperf]# /usr/local/bin/netperf -H 10.244.0.108 -l 5m -t UDP_STREAM -- -R 1 -m 1024
MIGRATED UDP STREAM TEST from 0.0.0.0 (0.0.0.0) port 0 AF_INET to 10.244.0.108 () port 0 AF_INET
send_data: data send error: No route to host (errno 113)
netperf: send_omni: send_data failed: No route to host
```

### After
`netperf` runs to completion.

```bash
[root@perf-client-other-node-749b79954c-hpgts netperf]# /usr/local/bin/netperf -H 10.244.0.108 -l 30s -t UDP_STREAM -- -R 1 -m 1024
MIGRATED UDP STREAM TEST from 0.0.0.0 (0.0.0.0) port 0 AF_INET to 10.244.0.108 () port 0 AF_INET
Socket  Message  Elapsed      Messages                
Size    Size     Time         Okay Errors   Throughput
bytes   bytes    secs            #      #   10^6bits/sec

212992    1024   30.00     1196041      0     326.60
212992           30.00     1072058            292.74
```

Fixes: #33159 

```release-note
wireguard: Fix issue where updates to a WireGuard device's configuration caused connectivity blips.
```

jrife committed:
docs: Add known issue for WireGuard connectivity issues

Signed-off-by: Jordan Rife <jrife@google.com>

jschwinger233 commented on pkg/wireguard/agent/agent.go:
```diff
@@ -572,15 +580,16 @@ func (a *Agent) OnIPIdentityCacheChange(modType ipcache.CacheModification, cidrC
 		if nodeName, ok := a.nodeNameByNodeIP[oldHostIP.String()]; ok {
 			if peer := a.peerByNodeName[nodeName]; peer != nil {
-				if peer.removeAllowedIP(ipnet) {
+				fmt.Println("Remove", ipnet)
```
Just for debugging?

jrife replied:
Oops must have missed this one.

jschwinger233 reviewed (commented):
Thanks! My understanding is we are maintaining a list of peer IPs that exists on wg dev. My concerns are:
1. This is pure userspace, so it could happen that a peer IP is removed from wg dev by command line, but cilium still thinks the IP is on dev.
2. We need synchronization to avoid racing of slice operations. Maybe there already is, if `UpdatePeer()` is called by `UpdateNode()` which is protected by a mutex, just saying we'd better double check.

jrife committed:
wireguard: Use dummy peer for allowed IP removal

Recent flakes in the "Network performance GKE" tests, in particular
those that use wireguard, uncovered an issue where updates to a peer's
allowed IPs can cause connectivity issues. In the best case this may
just result in dropped packets while in the worst case may lead to
failed calls to sendto/sendmsg (see GH-33159 for more detail on some of
the symptoms observed while running netperf).

This issue happens when ReplaceAllowedIPs is set, as it causes the
wireguard driver to completely clear the set of allowed IPs for a peer
before the set is rebuilt. This leaves a gap where packets sent or
received at that time will appear to not have a peer associated with
them. The current implementation sets ReplaceAllowedIPs every time
updatePeerByConfig is called, but we can avoid ever needing to use it.

This commit uses a "dummy peer" as a way to remove allowed IPs from a
peer rather than using ReplaceAllowedIPs. We use peerConfig to track any
pending IP inserts and removals and split these into two different sets
of updates in updatePeerByConfig. The first call to ConfigureDevice()
adds any IPs that are pending insertion. Next, any IPs that are pending
removal are moved to the "dummy peer" before removing the dummy peer
altogether, the net effect being that the IP is removed from the
original peer's list of allowed IPs.

This commit mostly sidesteps the issue that GH-34095 ("wireguard:
restore allowedIPs upong agent restart") set out to solve since we only
ever explicitly remove allowed IPs, meaning we can do away with a lot of
the accounting previously done by restoredPeers and restoredAllowedIPs.
However, we retain the behavior in RestoreFinished() that removes stale
peers and allowed IPs.

Fixes: #33159

Signed-off-by: Jordan Rife <jrife@google.com>

nathanjsweet reviewed (approved):
LGTM for CLI

gandro commented on pkg/wireguard/agent/agent.go:
```diff
@@ -732,97 +740,85 @@ type peerConfig struct {
+			delete(p.allowedIPs, pfx)
+			delete(p.needsRemove, pfx)
+		}
+	}
```
I wonder if it is worth setting the maps to `nil` if they are empty after we finished the sync. It will increase the number allocation/deallocations needed, but since otherwise the maps are never shrunk and basically consume memory forever.

jrife replied:
It probably wouldn't hurt. `UpdatePeer()` in particular has this part

```golang
peer.queueAllowedIPsInsert(a.ipCache.LookupByHostRLocked(nodeIPv4, nodeIPv6)...)
```

which would grow the capacity `needsInsert` well past what is needed later on. I do wonder about the effects of constant reallocation during periods of high pod churn.

gandro replied:
> I do wonder about the effects of constant reallocation during periods of high pod churn.

Yeah, it is a tradeoff. Ideally we'd cap the capacity of the map to a reasonable limit and reuse it as long as it's under the limit, and reset if if it's a above the limit.
Unfortunately it seems Golang has no way to inspect a map's capacity. We could track manually in a separate field which isn't pretty either. 

Not sure what to recommend here. I am worried that in particular with large-scale clustermesh setups, we might have up to 100k-300k of queued allowed IPs in total, meaning leaking about 10-30 MBs of memory.

(unrelated: we've also discussed adding pod CIDRs (where possihble) rather than individal pod IPs as allowedIPs, that would also reduce the pod churn issue. But that's not something we should tackle in this PR)

jrife replied:
> Unfortunately it seems Golang has no way to inspect a map's capacity.

You could potentially just use the map's length as a proxy for its capacity inside `finishAllowedIPSync()` to determine whether or not the capacity has reached a "high water mark". I think we should probably just start simple with setting the maps to `nil` every time here though. I think in the average case memory consumption is going to be the main thing we are concerned with.

jrife replied:
> unrelated: we've also discussed adding pod CIDRs (where possihble) rather than individal pod IPs as allowedIPs,

I was thinking about this as well. Are there any IPAM modes where there isn't a clear CIDR range per node where this approach wouldn't work? I suspect you may even get some performance benefit with WireGuard throughput this way, as it means quicker lookups in the driver's internal routing table to determine the peer associated with an IP. It would be interesting to test this out.

gandro replied:
> I was thinking about this as well. Are there any IPAM modes where there isn't a clear CIDR range per node where this approach wouldn't work? 

Yes, there are. ENI mode[^1], delegated IPAM mode, manual CRD mode, legacy Azure IPAM and AlibabaCloud all do not have the concept of a per-node pod CIDR. Instead, IPs are allocated from a global CIDR that is shared among all nodes (and pod IPs can be reassigned to other nodes). On such configurations, we will have to keep per-IP lists. 

But for IPAM modes `cluster-pool` (default), `kubernetes` (common, default on GKE) and `multi-pool`[^2] we could absolutely switch to a podCIDR based approach.

> It would be interesting to test this out.

That would be awesome. I'm more than happy to review such a PR.

[^1]: Some ENI nodes do run with prefix delegation, which means pod IPs are allocated out of /28 CIDRs on each node (with multiple CIDRs per node). That behaves a bit like pod CIDRs and would reduce the number of entries in "allowedIPs". However, those /28 CIDRs are not exposed to the agent's node manager at the moment, so there is no way for WireGuard to know them.
[^2]: `multi-pool` will have multiple CIDRs per node that are allocated dynamically, so it's a bit more involved than `cluster-pool` or `kubernetes` which have a single static pod CIDR. But still pretty much doable.

jrife replied:
> I think we should probably just start simple with setting the maps to nil every time here though.

Done in the latest push.

gandro commented on pkg/wireguard/agent/agent.go:
```diff
@@ -48,6 +46,7 @@ import (
 )
 
 var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "wireguard")
+var wgDummyPeerKey = wgtypes.Key{}
```
How does this look in `wg show`? I also wonder if we want to validate `UpsertPeer` to make sure no malicious CilumNode ever claims to be the dummy peer (which would probably break things).

jrife replied:
>  also wonder if we want to validate `UpsertPeer` to make sure no malicious CilumNode ever claims to be the dummy peer (which would probably break things).

Good point, I will add some validation logic around this.

jrife replied:
> How does this look in wg show? 

It's hard to capture with the PR as-is, since I delete the dummy peer immediately after moving IPs to it. For the sake of illustration, I commented that out locally to show what `wg show` would see. This shows before and after scaling a deployment down from 20 replicas to 1 replica. You can see that all the deleted IPs go to the dummy peer (`AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=`).

```
root@kind-worker:/# wg
interface: cilium_wg0
  public key: TnEOAoy+GCY3QMYiAUny3EpTT9DZ9VkW7KBUPLC03kU=
  private key: (hidden)
  listening port: 51871
  fwmark: 0x1e00

peer: ehHIOPWnKoJJUWBGlXQrokkGHpfu72Z6gqiNOuk8xBU=
  endpoint: 172.18.0.3:51871
  allowed ips: 172.18.0.3/32, 10.244.0.245/32, 10.244.0.60/32, fd00:10:244::a4c1/128, 10.244.0.32/32, 10.244.0.53/32, 10.244.0.115/32, 10.244.0.196/32, 10.244.0.84/32, 10.244.0.86/32, 10.244.0.81/32, 10.244.0.133/32, 10.244.0.171/32, 10.244.0.44/32, 10.244.0.186/32, 10.244.0.148/32, 10.244.0.83/32, 10.244.0.101/32, 10.244.0.230/32, 10.244.0.149/32, 10.244.0.92/32, 10.244.0.232/32, 10.244.0.207/32, 10.244.0.94/32, 10.244.0.135/32
root@kind-worker:/# wg
interface: cilium_wg0
  public key: TnEOAoy+GCY3QMYiAUny3EpTT9DZ9VkW7KBUPLC03kU=
  private key: (hidden)
  listening port: 51871
  fwmark: 0x1e00

peer: ehHIOPWnKoJJUWBGlXQrokkGHpfu72Z6gqiNOuk8xBU=
  endpoint: 172.18.0.3:51871
  allowed ips: 172.18.0.3/32, 10.244.0.245/32, 10.244.0.60/32, fd00:10:244::a4c1/128, 10.244.0.32/32, 10.244.0.53/32

peer: AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
  allowed ips: 10.244.0.101/32, 10.244.0.92/32, 10.244.0.196/32, 10.244.0.86/32, 10.244.0.81/32, 10.244.0.171/32, 10.244.0.44/32, 10.244.0.232/32, 10.244.0.148/32, 10.244.0.83/32, 10.244.0.186/32, 10.244.0.84/32, 10.244.0.207/32, 10.244.0.94/32, 10.244.0.149/32, 10.244.0.230/32, 10.244.0.135/32, 10.244.0.133/32, 10.244.0.115/32
```

jrife replied:
> I also wonder if we want to validate UpsertPeer to make sure no malicious CilumNode ever claims to be the dummy peer (which would probably break things).

I added a check in `UpdatePeer()` to make sure a node can't claim the dummy peer key.

gandro reviewed (commented):
I did a first pass and this is pretty clever and clean. I'll like do do a more thorough second review focusing a bit on the low-level bits later

gandro reviewed (approved):
Awesome work, this looks good to me now!
